ui/revenuecatui/src/test/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModelTest.kt

```diff
@@ -92,13 +105,159 @@ class PaywallViewModelTest {
     internal fun tearDown() {
         clearAllMocks()
     }
 
+    @Test
+    fun `Calls custom restore logic when purchasesAreCompletedBy == MY_APP`() = runTest {
+        every { purchases.purchasesAreCompletedBy } returns PurchasesAreCompletedBy.MY_APP
+
+        val myAppPurchaseLogic = mockk<MyAppPurchaseLogic>(relaxed = true)
+
+        coEvery { myAppPurchaseLogic.performRestore(any()) } returns MyAppRestoreResult.Success
+
+        val model = create(
+            customPurchaseLogic = myAppPurchaseLogic,
+            activeSubscriptions = setOf(TestData.Packages.monthly.product.id),
+            nonSubscriptionTransactionProductIdentifiers = setOf(TestData.Packages.lifetime.product.id)
+        )
+
+        model.awaitRestorePurchases()
+
+        coVerify { myAppPurchaseLogic.performRestore(customerInfo) }
+
+        verifyOrder {
+            listener.onRestoreStarted()
+            listener.onRestoreCompleted(customerInfo)
+        }
+    }
+
+    @Test
+    fun `Calls restore error listener when purchasesAreCompletedBy == MY_APP`() = runTest {
+        every { purchases.purchasesAreCompletedBy } returns PurchasesAreCompletedBy.MY_APP
+        val notAllowedError = PurchasesError(PurchasesErrorCode.PurchaseNotAllowedError)
+
+        val myAppPurchaseLogic = mockk<MyAppPurchaseLogic>(relaxed = true)
+
+        coEvery { myAppPurchaseLogic.performRestore(any()) } returns MyAppRestoreResult.Error(notAllowedError)
+
+        val model = create(
+            customPurchaseLogic = myAppPurchaseLogic,
+            activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
+        )
+
+        model.awaitRestorePurchases()
+
+        coVerify { myAppPurchaseLogic.performRestore(customerInfo) }
+
+        verifyOrder {
+            listener.onRestoreStarted()
+            listener.onRestoreError(notAllowedError)
+        }
+
+        assertThat(model.actionInProgress.value).isFalse
+    }
+
+    @Test
+    fun `Calls custom purchase logic when purchasesAreCompletedBy == MY_APP`() = runTest {
+        every { purchases.purchasesAreCompletedBy } returns PurchasesAreCompletedBy.MY_APP
+
+        val myAppPurchaseLogic = mockk<MyAppPurchaseLogic>(relaxed = true)
+
+        val purchase = stubGooglePurchase()
+        val myAppPurchase = MyAppPurchase(purchase)
+
+        coEvery { myAppPurchaseLogic.performPurchase(any(), any()) } returns MyAppPurchaseResult.Success(myAppPurchase)
+
+        val model = create(
+            customPurchaseLogic = myAppPurchaseLogic,
+            activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
+        )
+
+        model.awaitPurchaseSelectedPackage(activity)
+
+        coVerify { myAppPurchaseLogic.performPurchase(any(), any()) }
+
+        verifyOrder {
+            listener.onPurchaseStarted(any())
+            listener.onPurchaseCompleted(customerInfo, any())
+        }
+
+        assertThat(model.actionInProgress.value).isFalse
+        assertThat(dismissInvoked).isTrue
+    }
```

Comment:
We can modify this test, and tests like it, so we don't need the `await*()` call. 

1. Add an observable boolean using `MutableStateFlow`, initialized to `false`.
2. Construct a `MyAppPurchaseLogic` normally, without mocking. In `performPurchase`, set the observable boolean to `true`.
3. Call `purchaseSelectedPackage()` instead of `awaitPurchaseSelectedPackage()`.
4. Wait for the first emitted item from the observable boolean that is `true`. 

Altogether that makes it look something like this: 

```kotlin
fun `Calls custom purchase logic when purchasesAreCompletedBy == MY_APP`() = runTest {
    every { purchases.purchasesAreCompletedBy } returns PurchasesAreCompletedBy.MY_APP
    
    val purchase = stubGooglePurchase()
    val myAppPurchase = MyAppPurchase(purchase)
    
    val customPurchaseCalled = MutableStateFlow(false)
    val myAppPurchaseLogic = MyAppPurchaseLogic(
        performPurchase = { _, _ ->
            customPurchaseCalled.value = true
            MyAppPurchaseResult.Success(myAppPurchase)
        },
        performRestore = { error("Not expected to be called") }
    )
    
    val model = create(
        customPurchaseLogic = myAppPurchaseLogic,
        activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
    )
    
    model.purchaseSelectedPackage(activity)
    customPurchaseCalled.first { it }
    verifyOrder {
        listener.onPurchaseStarted(any())
        listener.onPurchaseCompleted(customerInfo, any())
    }
    
    assertThat(model.actionInProgress.value).isFalse
    assertThat(dismissInvoked).isTrue
}
```

Reply:
Interesting! `listener.onPurchaseCompleted(customerInfo, any())` is the last call to be made, after `performPurchase` has run - would it need to toggle a flag there instead?

Reply:
Ah yes! In that case a minimal diff from the existing test would be something like: 

```diff
@Test
fun `Calls custom purchase logic when purchasesAreCompletedBy == MY_APP`() = runTest {
+   val purchaseCompleted = MutableStateFlow(false)
    every { purchases.purchasesAreCompletedBy } returns PurchasesAreCompletedBy.MY_APP
+   every { listener.onPurchaseCompleted(any(), any()) }.answers {
+        purchaseCompleted.value = true
+   }

    val myAppPurchaseLogic = mockk<MyAppPurchaseLogic>(relaxed = true)

    val purchase = stubGooglePurchase()
    val myAppPurchase = MyAppPurchase(purchase)

    coEvery { myAppPurchaseLogic.performPurchase(any(), any()) } returns MyAppPurchaseResult.Success(myAppPurchase)

    val model = create(
        customPurchaseLogic = myAppPurchaseLogic,
        activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
    )

-    model.awaitPurchaseSelectedPackage(activity)
+    model.purchaseSelectedPackage(activity)
+    purchaseCompleted.first { it }

    coVerify { myAppPurchaseLogic.performPurchase(any(), any()) }

    verifyOrder {
        listener.onPurchaseStarted(any())
        listener.onPurchaseCompleted(customerInfo, any())
    }

    assertThat(model.actionInProgress.value).isFalse
    assertThat(dismissInvoked).isTrue
}
```

Reply:
That's very interesting, I was actually wondering how this test would look like without the await